feat: add `implicit` keyword and dedicated grammar

crusso opened the pull request:
Alternative to #5521 using `implicit` keyword to prefix argument declarations (rather than appearing in the type). 
Suggested by @christoph-dfinity.

Enforce restriction that `implicit` can only be used in argument positions, while preserving enough of the old syntax to have a non-breaking core.

* declare `implicit` as a keyword
* using dedicated argument patterns, with prefixed `implicit` keywords.
  
``` 
<implicit> ::= 
    'implicit'

<id_wild> ::=  
    '_'

<typ_item> ::= 
    <implicit> ':' <typ>
    <id> ':' <typ>
    <id_wild> ':' <typ>
    <implicit> <id> ':' <typ>
    <implicit> <id_wild> ':' <typ>

<pat_arg> ::= 
    <implicit> <id> ':' <typ>
    <implicit> <id> ':' <typ> '=' <pat>
    <implicit> <id_wild> ':' <typ> '=' <pat>
    ...

<pat_args> ::= 
    ...
    '(' <list(<pat_arg>, ',')> ')'

<func_pat> ::= 
    <id>? <typ_params_opt> <pat_args>
 ```
   
* allow `(implicit : typ)` named types, as before, to avoid breaking old clients of new core.

* rule out misplaced implicits in the type checker.  This turned out to be too difficult to do in the parser, though ugly to do in the checker.

Matches doc in #5674.

Example:

```motoko
module Array {
  public func toText<T>(as : [T], implicit toText : T -> Text) : Text {
     var t = "";
     for (a in as.vals()) {
       t := t # (toText(a));
     };
     t
  }
};
module Pair {
  public func toText<T,U>(
    implicit toText  :  T -> Text = toTextT,
    implicit toTextU : U -> Text = toTextU,
    p : (T, U))
    : Text {
      "(" # toTextT(p.0) # "," # toTextU(p.1) # ")"
    };
};

// types 
type F = <T, U>(implicit toText : T -> Text, implicit toText : U -> Text, p : (T, U)) -> Text;
```
[x] remove interim syntax
[x] rewrite tests
[x] add `implicit id : typ = pat` and `implicit  _ : typ = pat` (similar to field patterns) to support more syntaxtic manual override and wildcard pattern (without any new keyword).
[x] mo-doc works on core

crusso committed:
make arg to call a ref

crusso committed:
merge with master

crusso committed:
implement module (only) search

crusso committed:
plug holes

crusso committed:
test

crusso committed:
defer then check holes

crusso committed:
ca marche!

crusso committed:
extend test

crusso committed:
more tests

crusso committed:
update test output

crusso committed:
test dynamics

crusso committed:
match on names

crusso committed:
test local override

crusso committed:
extend test with redefinition of implicit

crusso committed:
improve error messages, add codes

crusso committed:
add implicit keyword and new grammar production to avoid parens

crusso committed:
missing file

crusso committed:
add repro

christoph-dfinity committed:
fixes for unary/nullary applications

crusso committed:
Merge branch 'claudio/implicits' into claudio/implicits-keyword

christoph-dfinity committed:
accept test output

christoph-dfinity committed:
fix: only check holes in non-pre phase

christoph-dfinity committed:
experiment: disambiguate's multiple matches by trying to find a single lower bound

christoph-dfinity committed:
only check inferred type for hole is closed in non-pre phase

crusso committed:
merge with parent

crusso committed:
Merge branch 'master' into claudio/implicits-keyword

crusso committed:
adjust tests

crusso committed:
Merge branch 'master' into claudio/implicits-keyword

crusso committed:
adjust tests

crusso committed:
fix test output with rebuild compiler

crusso committed:
fix grammar

crusso committed:
adjust viper test

crusso committed:
merge with master

crusso committed:
adjust test

crusso committed:
regen grammar.txt

crusso committed:
adjust grammar to accept legacy syntax too

crusso committed:
Merge branch 'master' into claudio/implicits-keyword

crusso committed:
allow prefix implicit

crusso committed:
add test

crusso committed:
revise example

crusso committed:
extend test to h.o

crusso committed:
update test output

crusso committed:
Merge branch 'master' into claudio/implicits-keyword

crusso committed:
Merge branch 'claudio/implicits-keyword' into claudio/implicits-keyword-alt

crusso committed:
distinguish argument patterns from other patterns, allowing implicits only in argument patterns; detect misplaced implicits in type as dynaic syntax errors

crusso committed:
update pretty printer and  test output

crusso committed:
Merge branch 'master' into claudio/implicits-keyword-redux

crusso committed:
update test output

crusso committed:
Merge branch 'claudio/implicits-keyword-redux' of github.com:dfinity/motoko into claudio/implicits-keyword-redux

crusso committed:
rewrite tests to new syntax

crusso committed:
check  placement during inference

crusso committed:
test error message

crusso committed:
fix tests

crusso committed:
cleanup

crusso committed:
Apply suggestion from @crusso

crusso committed:
Merge branch 'master' into claudio/implicits-keyword-redux

crusso committed:
tweak error message for implicit arguments

crusso committed:
fix bug with wild-card implicits, add test

crusso committed:
remove double quotation

crusso committed:
experiment: extend implicit patterns with rebinding and wildcard patterns (#5676)

Builds on #5659 to match doc in #5674.

Fills out argument pattern syntax to allow explicit renaming and
wildcards
```
<pat_arg> ::= 
    <implicit> <id> ':' <typ>
    <implicit> <id> ':' <typ> '=' <pat>
    <implicit> <id_wild> ':' <typ> '=' <pat>
    <pat_bin>
```

- [x] adjust pretty printer (sigh)

crusso committed:
Merge branch 'master' into claudio/implicits-keyword-redux

crusso committed:
adjust test output after merge with master

crusso committed:
Merge branch 'master' into claudio/implicits-keyword-redux

christoph-dfinity reviewed (approved):
Needs a changelog entry I guess? Unless we just make one big one for all the features combined.

crusso committed:
Merge branch 'master' into claudio/implicits-keyword-redux

crusso committed:
doc: implicit parameters (#5674)

Revised doc for dedicated `implicit (<id>| _) <typ>` syntax (including
wildcard patterns for now).

Matches #5659

---------

Co-authored-by: Christoph <christoph.hegemann@caffeine.ai>
